src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DefaultLSPProgressListener.cs

```diff
@@ -0,0 +1,192 @@
+            {
+                return;
+            }
+
+            var value = parameterToken[ProgressNotificationValueName];
+
+            try
+            {
+                request.HandlerCancellationToken.ThrowIfCancellationRequested();
+                await request.OnProgressNotifyAsync(value).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                // Initial handle request has been cancelled
+                // We can deem this ProgressRequest complete
+                ProgressCompleted(token);
+                return;
+            }
+
+            CompleteAfterDelay(token, request);
+        }
+
+        private void CompleteAfterDelay(string token, ProgressRequest request)
+        {
+            CancellationTokenSource linkedCTS;
+
+            lock (request.RequestLock)
+            {
+                var existingTimeoutCTS = request.TimeoutCancellationTokenSource;
+                if (existingTimeoutCTS != null &&
+                    existingTimeoutCTS.Token.CanBeCanceled &&
+                    !existingTimeoutCTS.Token.IsCancellationRequested)
+                {
+                    existingTimeoutCTS.Cancel();
+                }
+
+                request.TimeoutCancellationTokenSource = new CancellationTokenSource();
+                linkedCTS = CancellationTokenSource.CreateLinkedTokenSource(
+                    request.TimeoutCancellationTokenSource.Token,
+                    request.HandlerCancellationToken);
+            }
+
+            _ = CompleteAfterDelayAsync(token, request.TimeoutAfterLastNotify, linkedCTS.Token); // Fire and forget
+        }
+
+        private async Task CompleteAfterDelayAsync(string token, TimeSpan delay, CancellationToken cancellationToken)
+        {
+            try
+            {
+                await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                // Task cancelled, new progress notification received.
+                // Don't allow handler to return
+                return;
+            }
+
+            ProgressCompleted(token);
+        }
+
+        public override bool TryListenForProgress(
+            string token,
+            Func<JToken, Task> onProgressNotifyAsync,
+            TimeSpan timeoutAfterLastNotify,
+            CancellationToken handlerCancellationToken,
+            out Task onCompleted)
+        {
+            var onCompletedSource = new TaskCompletionSource<bool>();
+            var request = new ProgressRequest(
+                onProgressNotifyAsync,
+                timeoutAfterLastNotify,
+                handlerCancellationToken,
+                onCompletedSource);
+
+            if (!_activeRequests.TryAdd(token, request))
+            {
+                onCompleted = null;
+                return false;
+            }
+
+            CompleteAfterDelay(token, request);
```

Comment:
Start the delayed completion timer (if no FAR result is returned in 3 sec, return handler).

**Reasoning:**
Consider the platform cancels the original handler cancellation token before the first progress notification. In this case, the previous implementation would have never unblocked the handler.